### PR TITLE
make use of Context more consistent across functions and fn pointers

### PR DESCRIPTION
**Pull request by mhseiden**
This PR should have no functional impact. It just makes the use of `context.Context` more consistent throughout the codebase. That is, it is used as a fn argument wherever possible, and tries to use a user-supplied context where reasonable (rather than using `context.Background()` or `context.TODO()`).

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -170,11 +170,13 @@ func (sc *snowflakeConn) Close() (err error) {
 	sc.stopHeartBeat()
 
 	// ensure transaction is rollbacked
-	_, err = sc.exec(context.Background(), "ROLLBACK", false, false, nil)
+	ctx := context.TODO()
```
Comment (donhcd) [inline connection.go]: seems like maybe these methods should be taking contexts? but maybe a battle for another day

File: restful.go [hunk 2]
```diff
@@ -110,20 +110,15 @@ func postRestfulQuery(
 	data *execResponse, err error) {
 
 	requestID := uuid.New().String()
-
 	data, err = sr.FuncPostQueryHelper(ctx, sr, params, headers, body, timeout, requestID)
 
-	// errors other than context timeout and cancel would be returned to upper layers
-	if err != context.Canceled && err != context.DeadlineExceeded {
```
Comment (donhcd) [inline restful.go]: shouldn't we still have this line here? at least for this patch?

File: restful.go [hunk 3]
```diff
@@ -233,7 +231,7 @@ func closeSession(sr *snowflakeRestful) error {
 	headers["User-Agent"] = userAgent
 	headers[headerAuthorizationKey] = fmt.Sprintf(headerSnowflakeToken, sr.Token)
 
-	resp, err := sr.FuncPost(context.TODO(), sr, fullURL, headers, nil, 5*time.Second, false)
+	resp, err := sr.FuncPost(ctx, sr, fullURL, headers, nil, sr.RequestTimeout, false)
```
Comment (donhcd) [inline restful.go]: leave the timeout change for another patch?

Comment (mhseiden) [reply restful.go]: whoops, that was a bad diff

File: restful.go (hunk 2 above)
Comment (mhseiden) [reply restful.go]: Yep! Bad diff.

File: connection.go (hunk 1 above)
Comment (mhseiden) [reply connection.go]: These are the `database/sql` methods... which for some reason don't take a context. Seems like the fix here is to release local resources and then kick off a goroutine to close down the session on the snowflake side.

File: restful.go [hunk 4]
```diff
@@ -110,20 +110,17 @@ func postRestfulQuery(
 	data *execResponse, err error) {
 
 	requestID := uuid.New().String()
-
 	data, err = sr.FuncPostQueryHelper(ctx, sr, params, headers, body, timeout, requestID)
 
-	// errors other than context timeout and cancel would be returned to upper layers
 	if err != context.Canceled && err != context.DeadlineExceeded {
-		return data, err
+		// TODO this should not block the caller
```
Comment (donhcd) [inline restful.go]: shouldn't these be flipped?

Comment (mhseiden) [reply restful.go]: gha, you're right - thx
